### ISSUE TITLE
Null value for inversed one-to-one relation (v2.3.1)

### ISSUE THREAD

**Issue by florian-bailly** [zenstruck/foundry#749]
I'm upgrading Foundry on my project to use the v2.

However, it seems like the inversed relation is not auto filled, resulting in a null value (and violating the constraint on insertion), which was not the case before.

I thought [this issue](https://github.com/zenstruck/foundry/issues/655) would fix ours but it's not the case (since the latest version including the fix is not working).

For the reproducer, I created an Account and AccountSetting entity.

When trying to create an Account with a specified AccountSetting, the account property of the accountSetting is not filled (it is null).
```php
\App\Factory\AccountFactory::createOne([
  'accountSetting' => \App\Factory\AccountSettingFactory::new(),
]);
```

[Link to the reproducer test](https://github.com/florian-bailly/foundry-inversed-oneone-null/blob/main/tests/SampleTest.php)

**Comment by nikophil** [comment]
Please, could you try 2.x-dev and tell me how it's going?

**Comment by florian-bailly** [comment]
@nikophil on the dev branch, I get a duplicate key for my reproducer test BUT still a null value on project itself.

I can follow up the linked [issue](https://github.com/zenstruck/foundry/issues/752) if you prefer, and retest when a PR is available :)

**Comment by nikophil** [comment]
> I get a duplicate key for my reproducer test BUT still a null value on project itself.

Yeah, that's what I saw from your reproducer.

I've finally found a viable solution for this "one to one inversed" problem, just need to polish things up :) 
And I need to find why your case was not covered by our tests suite 🤔

>  I can follow up the linked issue if you prefer, and retest when a PR is available :)

Thanks, I'd rather to keep both issues open until I'm not 100% sure they are duplicates (we'll see if the patch fixes both issues!)

**Comment by nikophil** [comment]
@florian-bailly PR https://github.com/zenstruck/foundry/pull/755 fixes your issue, I just verified with your reproducer :)

**Comment by florian-bailly** [comment]
@nikophil indeed it's fixed for the reproducer case :)

However, I still have the null issue on my project on another entity. I'll try to make a reproducer for that case.

**Comment by nikophil** [comment]
yeah, if you can upgrade your reproducer that would be perfect :)

**Comment by javiereguiluz** [comment]
@nikophil in the docs, you show examples like this:

```php
$post = PostFactory::createOne(); // instance of Proxy

CommentFactory::createOne(['post' => $post]);
```

But what I want is to use just this:

```php
PostFactory::createOne();
```

And, inside the `PostFactory`, use something like this:

```php
protected function defaults(): array|callable
{
    return [
        // ...
        'comments' => [CommentFactory::createOne()],
    ];
}
```

This is not working as expected and in the error message I see something about null values for non-nullable fields.

I was going to create a new issue ... but then I found this issue and I don't know if it's the same issue reported here or not. Thanks!

**Comment by nikophil** [comment]
@javiereguiluz is the project where you have the problem open source? I'd really like to tackle this problem. More over your problem is about `OneToMany` relationship,  it's been the first time I hear of a problem with this kind of relation since a long time :thinking:

**Comment by florian-bailly** [comment]
@nikophil I think you have fixed the issue regarding the 1:1 since the issues I have are not for the same kind of relations.

I've updated the reproducer (sorry for the entity names, they are poorly chosen).

The error you should have is:
> An exception occurred while executing a query: SQLSTATE[08P01]: <<Unknown error>>: 7 ERROR:  bind message supplies 1 parameters, but prepared statement "" requires 2

and using xdebug I see that on the accountMultipleSettings table, the foo relation is not set... though in the test it's set explicitly.
<img width="627" alt="Screenshot 2024-12-24 at 14 49 40" src="https://github.com/user-attachments/assets/07ab92cf-5d8e-43c3-aee2-0b9b7dbe4472" />

Also, I'm wondering why the Foo entity is persisted twice. Shouldn't the `FooFactory::new()` in the AccountFactory defaults be overwritten by the attributes I set in the test?

Finally, how those regressions where possible between the v1 and the v2. Are there missing tests?

**Comment by nikophil** [comment]
> I've updated the reproducer

thanks, that will be super helpful

> I'm wondering why the Foo entity is persisted twice. Shouldn't the FooFactory::new() in the AccountFactory defaults be overwritten by the attributes I set in the test?

I still don't know, I have to deep dive into it :grimacing: 

> how those regressions where possible between the v1 and the v2. Are there missing tests?

yeah, there are literally tons of different use cases in the wild, and Foundry v2 got a nearly full rewrite
but our testsuite gets more and more robust these days :muscle: - by the way, I need to manage to reproduce your problem in Foundry's testsuite :grin:

**Comment by nikophil** [comment]
hi @florian-bailly 

this was challenging, but I finally tackled your problem!

By the way, still related to inversed one-to-one...  I don't know how Foundry 1 was actually better shaped for this kind of things :shrug:

**Comment by florian-bailly** [comment]
hi @nikophil ,

Well done!

I've tested the fix in my project, but I get the following error:
> Doctrine\ORM\ORMInvalidArgumentException: A new entity was found through the relationship 'Api\Entity\AccountSetting#account' that was not configured to cascade persist operations for entity: Api\Entity\Account@2273.
> To solve this issue: Either explicitly call EntityManager#persist() on this unknown entity or configure cascade persist this association in the mapping for example @ManyToOne(..,cascade={"persist"}).

Let me know if it's normal as the v2 is more strict or if you think it's a bug.

**Comment by nikophil** [comment]
gosh, this is a never ending story :disappointed: 

I cannot know whether the problem comes from Foundry or your code base only based on this error... 

The error is strange, because every entity created by Foundry is given to `$em->persist()`. But I cannot tell without any hint to reproduce the problem :shrug:

**Comment by florian-bailly** [comment]
@nikophil I managed to reproduce the error but after cleaning some files to make a simple reproducer, I came accross another error and I can't go back.

I updated the reproducer, hopefully after this last is fixed I'll be able to get back the one we're looking for :)

**Comment by nikophil** [comment]
hi @florian-bailly 

did you try [this PR](https://github.com/zenstruck/foundry/pull/775), which is not yet merged? it is the one which fixed your previous problem. Using it solves the problem in your reproducer :)

**Comment by florian-bailly** [comment]
@nikophil That's weird because I was on the PR 775 but it looks like a change was added after my reproducer.

Anyway, I confirm the case of the reproducer no 3 is fixed. BUT I still have the same error for another case (seems again a 1:1 with other relations).

I'll try to make a reproducer ;)

**Comment by nikophil** [comment]
Yeah, there were some changes on the PR indeed, but before I saw your new reproducer 😅
But I asked because the reproducer did not require the right branch 

Thanks for your help, though, very much appreciated 🙏

**Comment by nikophil** [comment]
hey @florian-bailly 

just waiting for https://github.com/zenstruck/foundry/pull/781to be reviewed/merged and I think all the bugs will be fixed 🎉 🤞

**Comment by florian-bailly** [comment]
hi @nikophil 

I confirm the PR 781 fixes the bug for the test 💪

However, I think there's a regression: I have a field A that is filled on PrePersist event using another field B which is set in the factory defaults. The error says that the field B cannot be accessed before initialization.
Let me know if you need a reproducer and another issue for that.

NB: I still have 15% of my test suite which is failing i.e. there are still cases to fix 🥺 I'll try to find some time to make reproducers, I've been quite busy lately.

**Comment by nikophil** [comment]
Just make the full testsuite public 😁

Yeah, let's close this issue if it is fixed, and open another one for another topic. As usual, at least provide a detailed description, and the best is always a public reproducer, or a failing test case in Foundry's  testsuite 😊

Out of a curiosity, how big is the testsuite we're talking about? Besides of Foundry's testsuite, before each release, I'm testing it on a big project which massively rely on foundry... But obviously that's not enough 😅

By the way, everything works better if you use `Factories` trait in every test using foundry (this will be enforced in a further version)

**Comment by nikophil** [comment]
I'm closing this issue. Don't hesitate to reopen another one for another problem 

> However, I think there's a regression: I have a field A that is filled on PrePersist event using another field B which is set in the factory defaults. The error says that the field B cannot be accessed before initialization.

I would be glad to have an example for this 🙏
